packages/es-components/src/components/containers/tabPanels/TabPanel.js

```diff
@@ -43,11 +43,19 @@ function TabPanel(props) {
     React.cloneElement(child, {
       selected: index === selectedIndex,
       action: header => {
-        const canChange = canTabChange(header);
-        if (canChange) {
-          setValue(header);
-          tabChanged(header);
+        const handleCanChange = (canChange) => {
+          if (canChange) {
+            setValue(header);
+            tabChanged(header);
+          }
+        };
+
+        const returnValue = canTabChange(header);
+        if (returnValue instanceof Promise) {
+          returnValue.then(handleCanChange);
+          return;
         }
+        handleCanChange(returnValue);
```

Comment:
In JS, you can actually `await` any method, even the ones that don't return a promise, so this function could basically change from what it was previously to:
```javascript
async header => {
  const canChange = await canTabChange(header);
  if (canChange) {
    setValue(header)
    tabChanged(header)
  }
}
```
That way we don't have to branch depending on if it's a Promise or not.

Reply:
We may also want to display some interstitial UI to show that we're waiting to determine if we can get to the next tab. But if your use case is anticipated to be extremely fast, we may not need that.

Reply:
Hmmm, my use case would conditionally overlay the embedded web app with a native app modal and the promise wouldn't resolve until the modal is closed. So, users wouldn't need the interstitial UI so much because they would have to interact with the native app instead. The conditions for showing the native modal _could_ involve network requests I suppose - if not now, then in the future. But, I haven't noticed the native app taking too long to determine whether the modal should be shown or not in my testing so far.

I think it brings up a good point too. Would async behavior be better implemented with new `checkingCanTabChange` & `canTabChangeComplete` sorts of props instead of `canTabChange`? This way consumers of the component can handle waiting as well.

Perhaps a larger discussion for handling async props as a whole is warranted. Would a code comment regarding interstitial UI suffice for now?

Reply:
P.S. Thanks for the async tip! That's nifty!

Reply:
Hey no problem!

I think, since we don't know if we'll need some kind of loading display right now, we should leave it out, and a comment would be fine.